library/lua/datetime.lua

```diff
@@ -0,0 +1,486 @@
+    return self:dayOfMonthTick() + 1, self:dayTick()
+end
+
+function DateTime:setDayOfMonth(day, month)
+    if not day or (day < 1 or day > self.calendar.daysPerMonth())
+        then qerror('day must be between 1 and %d, inclusive'):format(self.calendar.daysPerMonth())
+    end
+
+    if not month then month = self:month()
+    elseif (month < 1 or month > self.calendar.monthsPerYear()) then
+        qerror('month must be between 1 and %d, inclusive'):format(self.calendar.monthsPerYear())
+    end
+    -- zero based when converting to ticks
+    self.year_tick = self:monthsToTicks(month-1) + self:daysToTicks(day-1)
+    return self
+end
+
+-- returns a string in ordinal form (e.g. 1st, 12th, 22nd, 101st, 111th, 133rd)
+function DateTime:dayOfMonthWithSuffix()
+    local d = self:dayOfMonth()
+    return tostring(d)..getOrdinalSuffix(d)
+end
+
+-- returns an integer pair: (current month of the year, from 1), (month tick count)
+function DateTime:month()
+    return self:monthOfYearTick() + 1, self:monthTick()
+end
+
+function DateTime:setMonth(month)
+    if not month or (month < 1 or month > self.calendar.monthsPerYear()) then
+        qerror('month must be between 1 and %d, inclusive'):format(self.calendar.monthsPerYear())
+    end
+    self.year_tick = self:monthsToTicks(month-1)
+    return self
+end
+
+-- returns a string of the current month of the year
+function DateTime:nameOfMonth()
+    return self.calendar.months()[self:month()] or error("bad index?")
+end
+
+-- returns days since beginning of a year, starting from zero
+function DateTime:dayOfYearTick()
+    return self.year_tick // self.ticks_per_day
+end
+
+-- returns days since the beginning of a month, starting from zero
+function DateTime:dayOfMonthTick()
+    return self:dayOfYearTick() % self.calendar.daysPerMonth()
+end
+
+-- returns ticks since the beginning of a day
+function DateTime:dayTick()
+    return self.year_tick % self.ticks_per_day
+end
+
+-- sets the current day tick count
+function DateTime:setDayTick(ticks)
+    -- prevent rollover to next day
+    if not ticks or (ticks < 0 or ticks > self.ticks_per_day - 1) then
+        qerror('ticks must be between 0 and %d, inclusive'):format(self.ticks_per_day - 1)
+    end
+    self.year_tick = (self.year_tick - self:dayTick()) + ticks
+    return self
+end
+
+-- returns months since the beginning of a year, starting from zero
+function DateTime:monthOfYearTick()
+    return self.year_tick // self.ticks_per_month
+end
+
+-- returns ticks since the beginning of a month
+function DateTime:monthTick()
+    return self.year_tick % self.ticks_per_month
+end
+
+function DateTime:daysToTicks(days)
+    return days * self.ticks_per_day
+end
+
+function DateTime:monthsToTicks(months)
+    return months * self.ticks_per_month
+end
+
+function DateTime:addTicks(ticks)
+    if not ticks then qerror('ticks must be a number') end
+    self.year_tick = self.year_tick + ticks
+    self:normalize()
+    return self
+end
```

Comment:
I'm not sure how robust we want our argument checking to be. We could do:

`ticks = tonumber(ticks)
if not ticks...
`